### DISCUSSION TITLE
STORE_FAST__LOAD_FAST_SAME

### DISCUSSION THREAD

**Discussion by sweeneyde** [faster-cpython/ideas#299] [Ideas]
I mentioned this [here](https://github.com/faster-cpython/ideas/issues/16), but here's a possible super-instruction for when STORE_FAST is followed by a LOAD_FAST with the same oparg.

```c
        TARGET(STORE_FAST__LOAD_FAST_SAME) {
            PyObject *value = TOP();
            Py_INCREF(value);
            SETLOCAL(oparg, value);
            next_instr++;
            NOTRACE_DISPATCH();
        }
```

Benefits:
- no need to decode second oparg
- elide the pop-then-push stack manipulations
- no need to check for null in LOAD_FAST

I haven't benchmarked yet, but I have collected [stats for `python -m test` and stats for pyperformance](https://gist.github.com/sweeneyde/7d3414c6600500b43b5a41020cdfc4f4).
The test suite resulted in:

|Name | Count | Self |
|---|---:|---:|
| STORE_FAST__LOAD_FAST_SAME | 90557905 | 1.6% |
| STORE_FAST__LOAD_FAST | 87169944 | 1.6% |

And pyperformance resulted in

|Name | Count | Self |
|---|---:|---:|
| STORE_FAST__LOAD_FAST | 618849351 | 2.6% |
| STORE_FAST__LOAD_FAST_SAME | 413324475 | 1.7% |

This might also be beneficial in some comprehensions.

**Comment by brandtbucher** [comment]
Heh, I had the same idea and tried this a few weeks ago:

https://github.com/python/cpython/compare/main...brandtbucher:lf-lf

 I wasn't able to move the benchmarks, though (even when adding an additional `LOAD_FAST__LOAD_FAST_SAME` opcode on top of it).
